wgpu-core/src/device/mod.rs

```diff
@@ -556,6 +556,47 @@ impl<A: HalApi> Device<A> {
 
         let hal_usage = conv::map_texture_usage(desc.usage, desc.format.into());
 
+        let format_features = self
```

Comment:
can this check be moved out as well? This way, we can:
  1. make the function return `resource::Texture<A>` unconditionally
  2. don't pass `adapter` in

Reply:
yeah, we can move the check into `Global::create_texture_from_hal()`, and then replace `adapter` with `format_features` so it's not checked twice. see d38a400

If we're going to do that we could _also_ pass in `hal_usage` as a parameter instead of mapping from `desc.usage` twice.

---

But something tells me all this could be untangled more. I'd have to dig more into the types to see. 

Something that strikes me is that both `resource::Texture` and `hal::TextureDescriptor` need `hal_usage` (derived from `desc.usage` and `desc.format`). Especially when `resource::Texture` _already_ has access to `desc`.